### ISSUE TITLE
Incorrect workspace chat name after merging accounts

### ISSUE THREAD

**Issue by trjExpensify** [Expensify/App#14292]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Create a workspace from an account with a domain on the policyExpenseChat beta
2. Invite two members to the workspace, such that a workspace chat is created for each of invited members
3. Write a message in each of their workspace chats
4. Validate the first account on OldDot > sign-in on NewDot > account avatar > profile > display name > fill out first/last name
5. Write a message in the workspace chat > sign-out of the account
6. Sign in to newDot as the workspace admin > avatar > workspace > manage member > remove the first member from the workspace
7. Sign-out of the admin account
8. Validate the second account on OldDot > merge the old account into the new account
9. Sign-in to newDot as the second account
10. Observe that the %workspaceName% is now %memberOldPrimaryLogin%

## Expected Result:
- The workspace chat title from the member’s POV should always be the workspaceName, so the chat title should be: `workspaceName (archived)`
- From the admin’s POV the workspace chat title is always the memberName/memberPrimaryLogin (apart from their own workspace chat, of course)

## Actual Result:
- %workspaceName% is replaced with %memberOldPrimaryLogin% in the LHN and chat header.

## Workaround:
None

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [x] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

Note: I've only tested on these two, but I highly doubt it's not the case on the others. 

**Version Number:** v1.2.54-1
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 
<img width="1500" alt="image" src="https://user-images.githubusercontent.com/16232057/212382251-64f19a99-a22a-4184-999a-23e9e299d51b.png">

Here's are two comparisons the another archive cases (removing the member from the workspace, deleting the workspace) where it is displaying the workspaceName as the chat title correctly: 
<img width="1509" alt="image" src="https://user-images.githubusercontent.com/16232057/212382002-d684a379-f4a4-4e09-b7c1-a075af42f517.png">
<img width="1511" alt="image" src="https://user-images.githubusercontent.com/16232057/212382722-a24807a3-5e28-4dba-aa3b-77c5f1f31e9b.png">

**Expensify/Expensify Issue URL:** https://expensify.slack.com/archives/C02HWMSMZEC/p1673613908375389
**Issue reported by:** @trjExpensify 
**Slack conversation:**

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0174a9b9bdec5013d1</li>
        <li>Upwork Job ID: 1613952722062315520</li>
        <li>Last Price Increase: 2023-01-13</li>
    </ul>
</details>

**Comment by trjExpensify** [comment]
I've popped this issue on `internal` because it's behind a beta and I'm not sure yet whether it's purely a front-end issue or not. CC: @mountiny

**Comment by mountiny** [comment]
I will take this on @arosiclair if thats fine! 😄

**Comment by mountiny** [comment]
The workspace chat name is being decided in App [here](https://github.com/Expensify/App/blob/b9ee38a7d482c3fe015b3784490b137355f86869/src/libs/ReportUtils.js#L573-L576) and it is pending the `isOwnPolicyExpenseChat` report key.


That key is being set [here](https://github.com/Expensify/Web-Expensify/blob/77e8aaf9e56bec3e008281c8d8eacd7d9dbbc9e3/lib/GetAPI.php#L270) in Web-E and its true when the report `ownerEmail` is same as the user email. As part of the reportSummaryList.

This can be loaded through the GetChats command as well and in that case the property is defined here https://github.com/Expensify/Auth/blob/99a576a298748191f38acb6e0a13291b1054a277/auth/command/GetChats.cpp#L12

**Comment by mountiny** [comment]
This is most likely triggered form the AppInit call so that would be the GetChats, I will have to try to replicate locally so I can detect where the `isOwnPolicyExpenseChat ` is being set incorrectly.

**Comment by trjExpensify** [comment]
Ohh. I bet it's because when they merge accounts the ownerEmail is different? So we're viewing the workspace chat owned by the previous account (the one you merged from) as being "owned by someone else" from the perspective of the new account (the one you merged into)?

**Comment by mountiny** [comment]
That something I am suspecting, although I was not able to clearly identify where this happens in the backend, I think this should be transferred as well

**Comment by mountiny** [comment]
Havent got around to this one today with other issues, still aiming to explore this by the EOW

**Comment by mountiny** [comment]
Looked just for a bit, gonna be my priority coming forward

**Comment by mountiny** [comment]
Working on this today

**Comment by mountiny** [comment]
not much progress, continuing tomorrow

**Comment by mountiny** [comment]
Alright so:
- as userA merges old userB account into itself (userA), we pull all the userB policy expense chats (re-open the closed ones) so we can merge them and close again. We do it in the [archivePolicyExpenseChats](https://github.com/Expensify/Auth/blob/681770b427464d8dbfe2179b6403c80e25948f26/auth/lib/Account.cpp#L1447) method
- All the reports have [accountID of the old userB](https://github.com/Expensify/Auth/blob/681770b427464d8dbfe2179b6403c80e25948f26/auth/lib/Account.cpp#L1459)
- as we merge the accounts, we only [share](https://github.com/Expensify/Auth/blob/681770b427464d8dbfe2179b6403c80e25948f26/auth/lib/Account.cpp#L1486) the chat report which is archived with the new UserA
- Therefore the accountID at the report is never updated to the new account, it still points to the old account which was merged into userA

However, this issue is caused because of the accountID not being updated. 
- As user merges the old account to his account, the archived policy expense chat will have [ownerEmail](https://github.com/Expensify/Auth/blob/99a576a298748191f38acb6e0a13291b1054a277/auth/command/GetChats.cpp#L100) of the submitter for the report, which is still [the old account](https://github.com/Expensify/Auth/blob/99a576a298748191f38acb6e0a13291b1054a277/auth/command/GetChats.cpp#L77-L86) UserB in this case
- And finally as a consequence, this propagates all the way to NewDot, where the [ownerEmail](https://github.com/Expensify/App/blob/b9ee38a7d482c3fe015b3784490b137355f86869/src/libs/ReportUtils.js#L573-L576) of the chat is not same as the UserA and hence we show incorrect title of the archived/ merged chat.

How to solve this? Is it save to just update the accountID of the reports to point to the new account? I am not sure if there wont be any undesired side-effects here. 

I will ask in Slack.

**Comment by mountiny** [comment]
Asked here https://expensify.slack.com/archives/C02HWMSMZEC/p1674663453594989?thread_ts=1674536781.021579&cid=C02HWMSMZEC

**Comment by trjExpensify** [comment]
Okay cool, thanks for your efforts on this one so far Vit! Looks like we have a bit of git blame sleuthing to do as a next step to confirm this direction with the solution?

**Comment by mountiny** [comment]
> Looks like we have a bit of git blame sleuthing to do as a next step to confirm this direction with the solution?

Yep, I was looking into it and this part was introduced in [this PR](https://github.com/Expensify/Auth/pull/6455) and [original issue](https://github.com/Expensify/Expensify/issues/195475). However, I was not able to pinpoint any comment/ reason the policy expense chats were not updated with the new accountID.

I can try this locally and see if there is anything broken, but so far it seems like we should be able to do this.

**Comment by mountiny** [comment]
I have created a PR, I think there are some comments which point to why we have not been updating the accountIDs in a first place but I still have not wrapped my head around it. In the PR, we would update the accountIDs after we create a new policy expense chat for admin and share/unshare the old ones with the admin.

**Comment by trjExpensify** [comment]
Any clues in the associated detailed section in the doc, perhaps?

**Comment by mountiny** [comment]
concerns might be:
- Further, if  a user is re-added to the workspace, the workspace chat is [unarchived](https://expensify.slack.com/archives/C02HWMSMZEC/p1638295805063800). [Doc](https://docs.google.com/document/d/1OnaIAbVltGgdRWGdE53Y94qCKZxnW1DwDDarfz2jjMU/edit#bookmark=id.sh6u3aru5dg5)
    - I dont think this is a problem because if account is merged into another one, it cannot be accessed again with those login details, hence the reports could not be unarchived in this case
- Coming from [here in the doc](https://docs.google.com/document/d/1OnaIAbVltGgdRWGdE53Y94qCKZxnW1DwDDarfz2jjMU/edit#bookmark=id.pkq4ce19nuks), we have these options when merging accounts:
    - oldAccountID is an admin of the policy, and newAccountID is an admin of the policy - dont see a reason why owning the old admin's chats should be an issue
    - If oldAccountID is an admin of the policy, and newAccountID is NOT an admin of the policy - The old chats will all be archived and they cannot be un-archived again. 
    - If oldAccountID is an admin of the policy, and newAccountID is not part of the policy - the new account is made admin, no issue with switching the accountIDs
    - If oldAccountID is a member of the policy, and newAccountID is not a member of the policy - the new account is added, cant see an issue with owning it.
    
In general, dont see a problem here and the doc/ comments did not help me find out if there is any.

I am wondering, who is technically owner of the chat with all the admins?

**Comment by mountiny** [comment]
Adding reviewing label, but have to test more.

**Comment by mountiny** [comment]
I am having some issues with dev env which blocks me from testing it, trying to resolve it.

**Comment by mountiny** [comment]
Update is here https://expensify.slack.com/archives/C02HWMSMZEC/p1675120365788789?thread_ts=1674536781.021579&cid=C02HWMSMZEC

**Comment by mountiny** [comment]
Discussion still happening in slack

**Comment by mountiny** [comment]
Made some progress today

**Comment by mountiny** [comment]
Auth PR merged, App is in review.

**Comment by mountiny** [comment]
Done

**Comment by MelvinBot** [comment]
`Reviewing` label has been removed, please complete the "BugZero Checklist".

**Comment by MelvinBot** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.2.68-0 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/14779


If no regressions arise, payment will be issued on 2023-02-17. :confetti_ball:

After the hold period is over and BZ checklist items are completed, please complete any of the applicable payments for this issue, and check them off once done.
- [ ] External issue reporter
- [ ] Contributor that fixed the issue
- [ ] Contributor+ that helped on the issue and/or PR

As a reminder, here are the bonuses/penalties that should be applied for any External issue:
- Merged PR within 3 business days of assignment - 50% **bonus**
- Merged PR more than 9 business days after assignment - 50% **penalty**

**Comment by MelvinBot** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@mountiny] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@mountiny] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@mountiny] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@mountiny] Propose [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
- [ ] [@trjExpensify] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by trjExpensify** [comment]
Internal PRs and QAd right, so no payment is due here? We can close this out once the follow-up auth PR is in?

**Comment by mountiny** [comment]
Yeah no payments here, I need to:
-  remove the MERGED_0 from the actor name
- Check the flow when user has been member -> removed -> merged into an existing member

**Comment by trjExpensify** [comment]
Sounds good!

**Comment by MelvinBot** [comment]
@trjExpensify @mountiny this issue is now 4 weeks old and preventing us from maintaining WAQ. This should now be your highest priority. Please post below what your plan is to get a PR in review ASAP. Thanks!

**Comment by mountiny** [comment]
Its an evolving bug @MelvinBot, calm down buddy

**Comment by mountiny** [comment]
> remove the MERGED_0 from the actor name

Fixed this in App locally but I feel like this should be fixed in Web-E layer so I will look into that during the week

**Comment by mountiny** [comment]
Focusing on the manual requests doc, gonna try to work on this in spare time

**Comment by trjExpensify** [comment]
No change presumably, we're focused on getting manual requests out for detailed review.

**Comment by JmillsExpensify** [comment]
Yeah, sounds like we'll circle back on this next week once the detailed for Manual Requests is out. I think that's fine. This is one of the harder bugs to encounter. We'll close the loop well before other features in this newsletter.

**Comment by trjExpensify** [comment]
@mountiny we could to come back and close this out this week?

**Comment by mountiny** [comment]
Will try my best, coming back from ooo on wednesday and I will try to focus on this and manual requests

**Comment by mountiny** [comment]
Back and looking at this today

**Comment by trjExpensify** [comment]
Wahoo - let's finish this thing! ❤️

**Comment by mountiny** [comment]
PR which will fix the MERGED prefix in API layer is in a review.

We are left with the removed member merge flow

**Comment by mountiny** [comment]
figured why the closed account reason is wrong, as we switched to reportActionIDs instead of sequence numbers, this broke because we were not deleting the `removed from policy` closed report action and just adding another closed report action `merged accounts`.

With sequence numbers this worked because we showed the one which is the last one in the report. Now however this is random and seems like we are just showing the first CLOSED reportAction we run into and do not consider there could be more. Which makes sense as one would expect once the report is closed we wont add any report actions to it.

This is how the report actions look like in the report I tested locally, you can see there are two report actions.
```
sqlite> select * from reportActions where reportID=1551165745008980;
created                  reportID          accountID  action      message                                                                                               reportActionID
-----------------------  ----------------  ---------  ----------  ----------------------------------------------------------------------------------------------------  -------------------
2023-02-24 10:57:59.652  1551165745008980  195        CREATED                                                                                                           2501705207008724022
2023-02-24 11:02:28.364  1551165745008980  195        CLOSED      {"policyName":"MergeAccount","reason":"removedFromPolicy"}                                            2916847466558126623
2023-02-24 11:02:20.407  1551165745008980  187        ADDCOMMENT  {"html":"heya, you wont last long buddy"}                                                             4153088474101398725
2023-02-24 11:01:10.044  1551165745008980  195        ADDCOMMENT  {"html":"hello"}                                                                                      6575621110765283664
2023-02-24 11:19:02.690  1551165745008980  195        CLOSED      {"newLogin":"member+1@merge14779.com","oldLogin":"member+3@merge14779.com","reason":"accountMerged"}  7753211621765705462
2023-02-24 11:01:17.717  1551165745008980  195        ADDCOMMENT  {"html":"This is a message from member+3"}                                                            8914670120808483935
```

**Comment by trjExpensify** [comment]
Interesting! So do we need to update the logic to look for the latest closed reason then?

**Comment by mountiny** [comment]
> Interesting! So do we need to update the logic to look for the latest closed reason then?

Yep, looking into that now, I think that will be the simplest solution

**Comment by trjExpensify** [comment]
Sweet. Sounds good!

**Comment by mountiny** [comment]
App PR is up, filling out the screenshots.

**Comment by mountiny** [comment]
Web-E PR is merged so it should be in production on Tuesday.

App PR has been tested and reviewed and should be merged once Neil is online!

**Comment by trjExpensify** [comment]
Looking forward to running through the test scripts again!

**Comment by mountiny** [comment]
Merged the App PR 🎉

**Comment by MelvinBot** [comment]
`Reviewing` label has been removed, please complete the "BugZero Checklist".

**Comment by MelvinBot** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.2.78-0 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/15474


If no regressions arise, payment will be issued on 2023-03-13. :confetti_ball:

After the hold period is over and BZ checklist items are completed, please complete any of the applicable payments for this issue, and check them off once done.
- [ ] External issue reporter
- [ ] Contributor that fixed the issue
- [ ] Contributor+ that helped on the issue and/or PR

As a reminder, here are the bonuses/penalties that should be applied for any External issue:
- Merged PR within 3 business days of assignment - 50% **bonus**
- Merged PR more than 9 business days after assignment - 50% **penalty**

**Comment by MelvinBot** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@mountiny] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@mountiny] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@mountiny] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@mountiny] Propose [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
- [ ] [@trjExpensify] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by mountiny** [comment]
I think @trjExpensify you need to test this _again_

**Comment by trjExpensify** [comment]
Removing the payment hold etc. This was internally reported and handled in web/auth.

**Comment by trjExpensify** [comment]
On the agenda to test again this week.

**Comment by MelvinBot** [comment]
@trjExpensify, @mountiny Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by mountiny** [comment]
@trjExpensify given the tests discovered some polish issues handled separately:
- https://github.com/Expensify/App/issues/16324 
- https://github.com/Expensify/App/issues/16463

Both of them are reviewing, should we close this issue and do the final tests again once both of the fixes above are in staging?

**Comment by JmillsExpensify** [comment]
Given that we have the tracking issue, I think that's fine.

**Comment by trjExpensify** [comment]
Yep, I've added them to the OP of the tracker.
